File: 07/Java/TripleExpression/CheckedSqrt.java
package TripleExpression;

public class CheckedSqrt implements TripleExpression {
    private TripleExpression value;
    public CheckedSqrt(TripleExpression val) {
        value = val;
    }

    private void check(int a) throws Exception {
        if (a < 0) {
            throw new ParsingException("sqrt from negative number");
        }
    }

    public int evaluate(int x, int y, int z) throws Exception {
        int a = value.evaluate(x, y, z);
        check(a);
        int l = -1;
        int r = 46341;
        while (r - l > 1) {
            int m = (l + r) / 2;
            if (m*m <= a) {
                l = m;
            } else {
                r = m;
            }
        }
        return l;
    }
}

File: 07/CSharp/TripleExpressionApp/TripleExpressionApp/CheckedSqrt.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace TripleExpressionApp
{
    class CheckedSqrt : ITripleExpression
    {
        private ITripleExpression value;

        public CheckedSqrt(ITripleExpression value)
        {
            this.value = value;
        }

        protected void Check(double a)
        {
            if (a < 0)
            {
                throw new ParsingException("sqrt from negative number");
            }
        }

        public double Evaluate(double x, double y, double z)
        {
            double a = value.Evaluate(x, y, z);
            Check(a);
            double l = -1;
            double r = 46341;
            while (r - l > 1)
            {
                double m = (l + r) / 2;
                if (m * m <= a)
                {
                    l = m;
                }
                else
                {
                    r = m;
                }
            }
            return l;
        }
    }
}
